Tribler/Core/Libtorrent/LibtorrentMgr.py

```diff
@@ -631,9 +655,326 @@ def post_session_stats(self, hops=None):
+                       hidden=hidden or config.get_bootstrap_download())
+        return download
+
+    async def remove(self, download, removecontent=False, removestate=True):
+        infohash = download.get_def().get_infohash()
+        if infohash in self.downloads:
+            del self.downloads[infohash]
+        await download.stop(removestate=removestate, removecontent=removecontent)
+
+    def get_downloads(self):
+        return list(self.downloads.values())  # copy, is mutable
+
+    def get_download(self, infohash):
+        return self.downloads.get(infohash, None)
+
+    def get_channel_downloads(self):
+        return [download for download in self.downloads.values() if download.config.get_channel_download()]
+
+    def download_exists(self, infohash):
+        return infohash in self.downloads
+
+    async def update_download_hops(self, download, new_hops):
+        """
+        Update the amount of hops for a specified download. This can be done on runtime.
+        """
+        infohash = hexlify(download.tdef.get_infohash())
+        self._logger.info("Updating the amount of hops of download %s", infohash)
+        download.config.set_engineresumedata((await download.save_resume_data()))
+        await self.remove(download)
+
+        # copy the old download_config and change the hop count
+        config = download.config.copy()
+        config.set_hops(new_hops)
+        # If the user wants to change the hop count to 0, don't automatically bump this up to 1 anymore
+        config.set_safe_seeding(False)
+
+        self.add(download.tdef, config)
+
+    def update_trackers(self, infohash, trackers):
+        """ Update the trackers for a download.
+        :param infohash: infohash of the torrent that needs to be updated
+        :param trackers: A list of tracker urls.
+        """
+        dl = self.get_download(infohash)
+        old_def = dl.get_def() if dl else None
+
+        if old_def:
+            old_trackers = old_def.get_trackers_as_single_tuple()
+            new_trackers = list(set(trackers) - set(old_trackers))
+            all_trackers = list(old_trackers) + new_trackers
+
+            if new_trackers:
+                # Add new trackers to the download
+                dl.add_trackers(new_trackers)
+
+                # Create a new TorrentDef
+                if isinstance(old_def, TorrentDefNoMetainfo):
+                    new_def = TorrentDefNoMetainfo(old_def.get_infohash(), old_def.get_name(), dl.get_magnet_link())
+                else:
+                    metainfo = old_def.get_metainfo()
+                    if len(all_trackers) > 1:
+                        metainfo["announce-list"] = [all_trackers]
+                    else:
+                        metainfo["announce"] = all_trackers[0]
+                    new_def = TorrentDef.load_from_dict(metainfo)
+
+                # Set TorrentDef + checkpoint
+                dl.set_def(new_def)
+                dl.checkpoint()
+
+    # ???????
```

Comment:
Should this really be in devel?

Reply:
Oops, that's a leftover, gonna remove it...